Camel 16598 multicast blog (#575)
apache/camel-website: content/blog/2021/05/multicast-failure-routing/index.md

---
title: "Routing multicast output after encountering partial failures"
date: 2021-05-10
draft: false

> authors: [Reji Mathews, Hokutor]

Reviewer: These need to be the GitHub usernames, we use them to pull avatars from the GitHub profile, so you might want to set an image in your GitHub profile.

Author: Resolved . Falling back to github user id

Author: @zregvart - since its an array, could we add a co-author github ID too ?

Reviewer: Yeah, you can add as many as you need here, but they need to be GitHub usernames.

Author: Thanks. and appreciate your quick response :)

categories: ["Usecases"]
preview: "Routing multicast output after encountering partial failures"
---

## Problem description

> Multicast is a powerful EIP which supports parallel execution paths in asynchronous manner. There are various ways a camel user can configure a multicast EIP. Check out the extensive documentation [here](https://camel.apache.org/components/3.4.x/eips/multicast-eip.html)

Reviewer: ```suggestion
Multicast is a powerful EIP which supports parallel execution paths in asynchronous manner. There are various ways a camel user can configure a multicast EIP. Check out the extensive documentation [here](/components/latest/eips/multicast-eip.html)
```

Let's use the latest version and we need to use relative links for the link checker to check links to camel.apache.org.

1. One can configure to execute all the child paths independently and continue routing the last reply as the outgoing message (default behavior unless you provide an aggregation strategy)

> 2. Additionally, you can plug in an implementation of [camel aggregation strategy]( https://github.com/apache/camel/blob/camel-3.7.x/core/camel-api/src/main/java/org/apache/camel/AggregationStrategy.java ) with user defined logic to aggregate the output from each of those child path before continuing further downstream routing. 

Reviewer: ```suggestion
2. Additionally, you can plug in an implementation of [camel aggregation strategy](https://github.com/apache/camel/blob/main/core/camel-api/src/main/java/org/apache/camel/AggregationStrategy.java) with user defined logic to aggregate the output from each of those child path before continuing further downstream routing. 
```

Perhaps also point to the `main` branch?

2021-05-06 12:43:19.575  INFO 13956 --- [ #3 - Multicast] route3                                   : Executing PATH_2 - success path
2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] route3                                   : PATH_2
2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator DATA_FROM_PATH_2
```

## What could take you by a surprise?

- When the multicast completes aggregating exchanges from child branches, one might intermittently note that it stops routing the remaining processors (those 2 additional log and a transform step in above example). On execution tracing exercise, you will notice this happens in a special scenario when the very first exchange which arrives in the aggregator (from first completed child branch) had encountered an exception during its course or/and was handled via onException flows. On the flip side, if the first exchange was a successful and even though all the remaining ones experienced a failure, it still continued routing the remaining processors/steps.

## Analysis

> To understand this better, lets deep dive into the open source codebase. Notice the [Pipeline class]( https://github.com/apache/camel/blob/6dff85675e9b73e8a528bc2683935ec3c1ed26b7/core/camel-core-processor/src/main/java/org/apache/camel/processor/Pipeline.java ). PipelineProcessor (part of camel core framework processors) performs an evaluation after every user processors (user added steps in a camel flow) on whether it should continue routing to the next processor.

Reviewer: ```suggestion
To understand this better, lets deep dive into the open source codebase. Notice the [Pipeline class](https://github.com/apache/camel/blob/main/core/camel-core-processor/src/main/java/org/apache/camel/processor/Pipeline.java). PipelineProcessor (part of camel core framework processors) performs an evaluation after every user processors (user added steps in a camel flow) on whether it should continue routing to the next processor.
```

Would it make sense to point to `main` branch?

> This decision is made inside Pipeline class on [this if block](https://github.com/apache/camel/blob/6dff85675e9b73e8a528bc2683935ec3c1ed26b7/core/camel-core-processor/src/main/java/org/apache/camel/processor/Pipeline.java#L79) <br>

Reviewer: Yeah I see the issue, the `main` branch could diverge and this line would not be valid any more. Do you think it makes sense to provide a snippet of the code from `Pipeline.java` here instead?

Author: Yes makes sense

Author: done. can you review now?

---
title: "Routing multicast output after encountering partial failures"
date: 2021-05-10
draft: false
authors: [hokutor, mathewsreji]
categories: ["Usecases"]
preview: "Routing multicast output after encountering partial failures"
---

## Problem description

> Multicast is a powerful EIP which supports parallel execution paths in asynchronous manner. There are various ways a camel user can configure a multicast EIP. Check out the extensive documentation [here](/components/latest/eips/multicast-eip.html)

Reviewer: ```suggestion
Multicast is a powerful EIP which supports parallel execution paths in asynchronous manner. There are various ways a Camel user can configure a multicast EIP. Check out the extensive documentation [here](/components/latest/eips/multicast-eip.html)
```

> Multicast is a powerful EIP which supports parallel execution paths in asynchronous manner. There are various ways a camel user can configure a multicast EIP. Check out the extensive documentation [here](/components/latest/eips/multicast-eip.html)

Reviewer: ```suggestion
Multicast is a powerful EIP that asynchronously supports parallel execution paths. There are various ways a camel user can configure a multicast EIP. Check out the extensive documentation [here](/components/latest/eips/multicast-eip.html)
```

1. One can configure to execute all the child paths independently and continue routing the last reply as the outgoing message (default behavior unless you provide an aggregation strategy)

> 2. Additionally, you can plug in an implementation of [camel aggregation strategy](https://github.com/apache/camel/blob/main/core/camel-api/src/main/java/org/apache/camel/AggregationStrategy.java) with user defined logic to aggregate the output from each of those child path before continuing further downstream routing. 

Reviewer: ```suggestion
2. Additionally, you can plug in an implementation of a [Camel aggregation strategy](https://github.com/apache/camel/blob/main/core/camel-api/src/main/java/org/apache/camel/AggregationStrategy.java) with user-defined logic to aggregate the output from each of those child paths before continuing further downstream routing. 
```

> For the use case discussed below, the requirement is to aggregate the computed results from all child paths before it gets routed to the downstream processors in the flow. The idea is to keep routing the aggregated results if atleast one child route completes successfully without an exception. We also want to stop routing further if all the child exchanges experienced failures.

Reviewer: ```suggestion
For the use case discussed below, the requirement is to aggregate the computed results from all child paths before it gets routed to the downstream processors in the flow. The idea is to keep routing the aggregated results if at least one child route completes successfully without an exception. We also want to stop routing further if all the child exchanges experienced failures.
```

## Use case 

> Check out the following camel routes 

Reviewer: ```suggestion
Check out the following Camel routes 
```

```log
2021-05-06 12:43:18.565  INFO 13956 --- [qtp916897446-42] route1                                   : received request
2021-05-06 12:43:18.566  INFO 13956 --- [qtp916897446-42] route1                                   : Entering multicast
2021-05-06 12:43:18.575  INFO 13956 --- [ #4 - Multicast] route2                                   : Executing PATH_1 - exception path
2021-05-06 12:43:18.575  INFO 13956 --- [ #4 - Multicast] route2                                   : Starting exception throw
2021-05-06 12:43:18.578  INFO 13956 --- [ #4 - Multicast] route2                                   : Exception handler invoked
2021-05-06 12:43:18.579  INFO 13956 --- [ #4 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator {"data" : "err"}
2021-05-06 12:43:19.575  INFO 13956 --- [ #3 - Multicast] route3                                   : Executing PATH_2 - success path
2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] route3                                   : PATH_2
2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator DATA_FROM_PATH_2
```

## What could take you by a surprise?

> - When the multicast completes aggregating exchanges from child branches, one might intermittently note that it stops routing the remaining processors (those 2 additional log and a transform step in above example). On execution tracing exercise, you will notice this happens in a special scenario when the very first exchange which arrives in the aggregator (from first completed child branch) had encountered an exception during its course or/and was handled via onException flows. On the flip side, if the first exchange was a successful and even though all the remaining ones experienced a failure, it still continued routing the remaining processors/steps.

Reviewer: ```suggestion
- When the multicast completes aggregating exchanges from child branches, one might intermittently note that it stops routing the remaining processors (those two additional log and a transform steps in the above example). On execution tracing exercise, you will notice this happens in a special scenario when the very first exchange which arrives in the aggregator (from the first completed child branch) had encountered an exception during its course or/and was handled via `onException` flows. On the flip side, if the first exchange was successful and even though all the remaining ones experienced a failure, it continued routing the remaining processors/steps.
```

## Analysis

> To understand this better, lets deep dive into the open source codebase. Check out PipelineProcessor.java (part of camel-core-processors module). Following section of code in the class Pipeline performs an evaluation after every user processors (user added steps in a camel flow) on whether it should continue routing to the next processor.

Reviewer: ```suggestion
To understand this better let's deep dive into the open source codebase. Check out `PipelineProcessor.java` (part of `camel-core-processors` module). The following section of code in the class `Pipeline` evaluates after every user processor (user added steps in a Camel flow) on whether it should continue routing to the next processor.
```

                || (ee.isErrorHandlerHandledSet() && ee.isErrorHandlerHandled());
        if (stop) {
            if (log.isDebugEnabled()) {
                StringBuilder sb = new StringBuilder();
                sb.append("Message exchange has failed: ").append(message).append(" for exchange: ").append(exchange);
                if (exchange.isRollbackOnly() || exchange.isRollbackOnlyLast()) {
                    sb.append(" Marked as rollback only.");
                }
                if (exchange.getException() != null) {
                    sb.append(" Exception: ").append(exchange.getException());
                }
                if (ee.isErrorHandlerHandledSet() && ee.isErrorHandlerHandled()) {
                    sb.append(" Handled by the error handler.");
                }
                log.debug(sb.toString());
            }

            return false;
        }
        if (ee.isRouteStop()) {
            if (log.isDebugEnabled()) {
                log.debug("ExchangeId: {} is marked to stop routing: {}", exchange.getExchangeId(), exchange);
            }
            return false;
        }

        return true;
    }

}
```
 

> Well, now lets re-visit our use case. What if you still want to continue routing?

Reviewer: ```suggestion
Well, now let's re-visit our use case. What if you still want to continue routing?
```

> - From our above aggregator, you will notice that the very first exchange which arrives in aggregator becomes the base exchange on which the aggregator continues to pile up body content (with incoming results from other child routes). Infact, a lot of camel users follow this pattern of writing an aggregator strategy. Unfortunately, if done this way, the state variables set on the Exchange object during exception handling get carried forward to the next evaluation point in Pipeline and qualify to stop routing. 

Reviewer: ```suggestion
- From our above aggregator, you will notice that the very first exchange which arrives in aggregator becomes the base exchange on which the aggregator continues to pile up body content (with incoming results from other child routes). In fact, a lot of Camel users follow this pattern of writing an aggregator strategy. Unfortunately, if done this way, the state variables set on the Exchange object during exception handling get carried forward to the next evaluation point in Pipeline and qualify to stop routing. 
```

> - From our above aggregator, you will notice that the very first exchange which arrives in aggregator becomes the base exchange on which the aggregator continues to pile up body content (with incoming results from other child routes). Infact, a lot of camel users follow this pattern of writing an aggregator strategy. Unfortunately, if done this way, the state variables set on the Exchange object during exception handling get carried forward to the next evaluation point in Pipeline and qualify to stop routing. 

Reviewer: ```suggestion
- From our above aggregator, you will notice that the very first exchange that arrives in the aggregator becomes the base exchange on which the aggregator continues to accumulate body contents (with incoming results from other child routes). In fact, a lot of Camel users follow this pattern of writing an aggregator strategy. Unfortunately, if done this way, the state variables set on the Exchange object during exception handling get carried forward to the next evaluation point in Pipeline and qualify to stop routing. 
```

<br>

## Solution 

There are many ways a user could neutralize the states set by the exception handling framework. However, for the scope of this article, we chose the following strategy. 

> - If first child route exchange never encountered an exception, then continue processing the rest of aggregation cycle as usual.

Reviewer: ```suggestion
- If the first child route exchange never encountered an exception, then continue processing the rest of the aggregation cycle as usual.
```

> - If the first child encountered an excetion, then introspect the incoming exchanges for success case. If found, shift the base to be the first succeess exchange and move the aggregated results on to it and continue the rest of aggregation lifecycle as usual.

Reviewer: ```suggestion
- If the first child encountered an exception, then introspect the incoming exchanges for success case. If found, shift the base to be the first successful exchange and move the aggregated results on to it and continue the rest of the aggregation lifecycle as usual.
```

            if(!hadException(newExchange)) {
                // aggregate and swap the base
                LOGGER.info("Found new exchange with success. swapping the base exchange");
                List<String> oldData = oldExchange.getIn().getBody(List.class);
                oldData.add(newExchange.getIn().getBody(String.class));
                newExchange.getIn().setBody(oldData); // swapped the base here
                return newExchange;
            }
        }

        List<String> oldData = oldExchange.getIn().getBody(List.class);
        oldData.add(newExchange.getIn().getBody(String.class));
        oldExchange.getIn().setBody(oldData);

        return oldExchange;
    }

    private boolean hadException(Exchange exchange) {

        if(exchange.isFailed()) {
            return true;
        }

        if(exchange.isRollbackOnly()) {
            return true;
        }

        if(exchange.isRollbackOnlyLast()) {
            return true;
        }

        if(((ExtendedExchange)exchange).isErrorHandlerHandledSet()
                && ((ExtendedExchange)exchange).isErrorHandlerHandled()) {
            return true;
        }

        return false;
    }
}

> ````

Reviewer: There is one extra backtick here and I can't make a suggestion for it, we need three not four here.

Author: Fixed this too

2021-05-06 12:46:19.134  INFO 2576 --- [ #3 - Multicast] route2                                   : Exception handler invoked
2021-05-06 12:46:19.135  INFO 2576 --- [ #3 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator {"data" : "err"}
2021-05-06 12:46:20.130  INFO 2576 --- [ #4 - Multicast] route3                                   : Executing PATH_2 - success path
2021-05-06 12:46:22.132  INFO 2576 --- [ #4 - Multicast] route3                                   : PATH_2
2021-05-06 12:46:22.132  INFO 2576 --- [ #4 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator DATA_FROM_PATH_2
2021-05-06 12:46:22.132  INFO 2576 --- [ #4 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Found new exchange with success. swapping the base exchange
2021-05-06 12:46:22.133  INFO 2576 --- [ #4 - Multicast] route1                                   : Aggregated results {"data" : "err"},DATA_FROM_PATH_2
2021-05-06 12:46:22.133  INFO 2576 --- [ #4 - Multicast] route1                                   : Another log
```

With the new aggregator implementation, you can now see the exchange getting routed down to the remaining processors.

> Hope this article helps users who hit upon this issue. 

Reviewer: ```suggestion
Hope this article helps users who encounter this issue.
```
